File: chat-study-auth/src/main/java/com/example/config/DataSourceConfig.java
package com.example.config;

import com.example.database.RoutingDataSource;
import com.zaxxer.hikari.HikariDataSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.jdbc.DataSourceBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;
import org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.SQLException;
import java.util.HashMap;
import java.util.Map;

@Configuration
public class DataSourceConfig {
    private static final Logger log = LoggerFactory.getLogger(DataSourceConfig.class);

    @ConfigurationProperties(prefix = "spring.datasource.source.hikari")
    @Bean
    public DataSource sourceDataSource() {
        return DataSourceBuilder.create().type(HikariDataSource.class).build();
    }

    @ConfigurationProperties(prefix = "spring.datasource.replica.hikari")
    @Bean
    public DataSource replicaDataSource() {
        return DataSourceBuilder.create().type(HikariDataSource.class).build();
    }

    @Bean
    public DataSource routingDataSource(@Qualifier("sourceDataSource") DataSource sourceDataSource, @Qualifier("replicaDataSource") DataSource replicaDataSource) throws SQLException {
        RoutingDataSource routingDataSource = new RoutingDataSource();

        Map<Object, Object> targetDataSources = new HashMap<>();
        targetDataSources.put("source", sourceDataSource);
        targetDataSources.put("replica", replicaDataSource);

        routingDataSource.setTargetDataSources(targetDataSources);
        routingDataSource.setDefaultTargetDataSource(sourceDataSource);

        try (Connection connection = replicaDataSource.getConnection()) {
            log.info("Init ReplicaConnectionPool.");
        }

        return routingDataSource;
    }

    @Primary
    @Bean
    public DataSource lazyConnectionDataSource(@Qualifier("routingDataSource") DataSource routingDataSource) {
        return new LazyConnectionDataSourceProxy(routingDataSource);
    }
}

File: loop-pack-be-l2-vol3-kotlin/modules/jpa/src/main/kotlin/com/loopers/config/jpa/DataSourceConfig.kt
package com.loopers.config.jpa

import com.zaxxer.hikari.HikariConfig
import com.zaxxer.hikari.HikariDataSource
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary

@Configuration
class DataSourceConfig {
    @Bean
    @ConfigurationProperties(prefix = "datasource.mysql-jpa.main")
    fun mySqlMainHikariConfig(): HikariConfig =
        HikariConfig()

    @Primary
    @Bean
    fun mySqlMainDataSource(@Qualifier("mySqlMainHikariConfig") hikariConfig: HikariConfig) =
        HikariDataSource(hikariConfig)
}
